packages/superset-ui-connection/src/SupersetClient.ts

```diff
@@ -0,0 +1,200 @@
+    this.csrfPromise = Promise.reject({
+      error: `SupersetClient has no CSRF token, ensure it is initialized or
+      try logging into the Superset instance at ${this.getUrl({
+        host: this.host,
+        endpoint: '/login',
+      })}`,
+    });
+
+    if (typeof this.csrfToken === 'string') {
+      this.headers = { ...this.headers, 'X-CSRFToken': this.csrfToken };
+      this.csrfPromise = Promise.resolve(this.csrfToken);
+    }
+  }
+
+  public async get({
+    body,
+    credentials,
+    headers,
+    host,
+    endpoint,
+    mode,
+    parseMethod,
+    signal,
+    timeout,
+    url,
+  }: RequestConfig): Promise<Response | JsonResponse | TextResponse> {
+    return this.ensureAuth().then(() =>
+      callApi({
+        body,
+        credentials: credentials || this.credentials,
+        headers: { ...this.headers, ...headers },
+        method: 'GET',
+        mode: mode || this.mode,
+        parseMethod,
+        signal,
+        timeout: timeout || this.timeout,
+        url: this.getUrl({ endpoint, host: host || this.host, url }),
+      }),
+    );
+  }
+
+  public async post({
+    credentials,
+    endpoint,
+    headers,
+    host,
+    mode,
+    parseMethod,
+    postPayload,
+    signal,
+    stringify,
+    timeout,
+    url,
+  }: RequestConfig): Promise<Response | JsonResponse | TextResponse> {
+    return this.ensureAuth().then(() =>
+      callApi({
+        credentials: credentials || this.credentials,
+        headers: { ...this.headers, ...headers },
+        method: 'POST',
+        mode: mode || this.mode,
+        parseMethod,
+        postPayload,
+        signal,
+        stringify,
+        timeout: timeout || this.timeout,
+        url: this.getUrl({ endpoint, host: host || this.host, url }),
+      }),
+    );
+  }
+
+  private ensureAuth() {
+    return this.csrfPromise;
+  }
+
+  public isAuthenticated(): boolean {
+    // if CSRF protection is disabled in the Superset app, the token may be an empty string
+    return this.csrfToken !== null && this.csrfToken !== undefined;
+  }
+
+  public init(force: boolean = false) {
+    if (this.isAuthenticated() && !force) {
+      return this.csrfPromise;
+    }
+
+    return this.getCSRFToken();
+  }
+
+  private async getCSRFToken() {
+    this.csrfToken = undefined;
+
+    // If we can request this resource successfully, it means that the user has
+    // authenticated. If not we throw an error prompting to authenticate.
+    this.csrfPromise = callApi({
+      credentials: this.credentials,
+      headers: {
+        ...this.headers,
+      },
+      method: 'GET',
+      mode: this.mode,
+      timeout: this.timeout,
+      url: this.getUrl({ endpoint: 'superset/csrf_token/', host: this.host }),
+    }).then((response: JsonResponse | any) => {
+      if (response.json) {
+        this.csrfToken = response.json.csrf_token;
+        if (typeof this.csrfToken === 'string') {
+          this.headers = { ...this.headers, 'X-CSRFToken': this.csrfToken };
```

Comment:
Adding some types to guarantee that `headers` includes the right properties would probably be pretty helpful. Are there certain headers that are always passed on every request? Maybe something like:
```typescript
type RequiredHeaders = {
  'X-CSRFToken': string;
  // ...
};
type Headers = { [k: string]: string } & RequiredHeaders;
```

One nice side-effect of this is that there's less of a need for the `csrfToken` type alias. As long as the type for the final `headers` object that gets passed into `fetch` is correct, TS can guarantee that all the types that might populate that field are correct. Then it doesn't require as much care to make sure that `crsfToken` is always the same type for all of the request types.

Reply:
only `'X-CSRFToken'`is required for now, but this a great idea, thanks for the suggestion!

Reply:
For now I think we'll punt on required headers vs not, this is a pretty generic client for the time being.